larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+        ClusterList threeDClusterList;
+        LArPfoHelper::GetThreeDClusterList(pInputPfo, threeDClusterList);
+
+        if (threeDClusterList.empty())
+            return;
+
+        CaloHitList threeDCaloHitList; // Might be able to get through LArPfoHelper instead
+        const Cluster *const pThreeDCluster(threeDClusterList.front());
+        pThreeDCluster->GetOrderedCaloHitList().FillCaloHitList(threeDCaloHitList);
+
+        if (threeDCaloHitList.empty())
+            return;
+
+        // Run the PCA analysis
+        CartesianVector centroid(0.f, 0.f, 0.f);
+        LArPcaHelper::EigenVectors eigenVecs;
+        LArPcaHelper::EigenValues eigenValues(0.f, 0.f, 0.f);
+        LArPcaHelper::RunPca(threeDCaloHitList, centroid, eigenValues, eigenVecs);
+        const float principalEigenvalue(eigenValues.GetX()), secondaryEigenvalue(eigenValues.GetY()), tertiaryEigenvalue(eigenValues.GetZ());
+
+        if (m_ratioVariables)
+        {
+            float pca1(-1.f), pca2(-1.f);
+            if(principalEigenvalue > std::numeric_limits<float>::epsilon())
+            {
+                pca1 = secondaryEigenvalue/principalEigenvalue;
+                pca2 = tertiaryEigenvalue/principalEigenvalue;
+            }
+            featureVector.push_back(pca1);
+            featureVector.push_back(pca2);
+        }
+        else
+        {
+            featureVector.push_back(principalEigenvalue);
+            featureVector.push_back(secondaryEigenvalue);
+            featureVector.push_back(tertiaryEigenvalue);
+        }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode PCAFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+{
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "RatioVariables", m_ratioVariables));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ChargeFeatureTool::ChargeFeatureTool() :
+    m_ratioVariables(true),
+    m_addTotalCharge(false),
+    m_addChargeSigma(true),
+    m_addChargeIncrements(false),
+    m_addStartCharge(false),
+    m_addEndCharge(true),
+    m_endChargeFraction(0.1f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ChargeFeatureTool::CalculateChargeVariables(const Algorithm *const pAlgorithm, const pandora::Cluster *const pCluster, float &totalCharge, float &chargeSigma, 
+    float &chargeMean, float &startCharge, float &endCharge, float &incrementCharge, float endChargeFraction)
```

Comment:
Might be better to create a little class to hold all these variables, with a default constructor setting them all to their initial values as appropriate. Then just have a single argument, by non-const reference, to pass in and receive populated ChargeResults.

Reply:
could be done that way, but following a comment I made later on, maybe it's better to remove the features I didn't use in the end? In that way we would only have two variables filled and would simplify this a lot. I added (almost) all the different features I tested, but maybe it's not really interesting other than for myself in development and I could just remove some stuff for simplicity, thoughtS?